### PR TITLE
[NNC] refactor cuda half support to more general file

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47374 [NNC] Enable unary op cpu testing
* **#47373 [NNC] refactor cuda half support to more general file**
* #47372 [NNC] More cpu tests
* #47371 [NNC] Add more CPU Tests
* #47370 [NNC] Fix llvm min lowering for int inputs
* #47369 Add more CPU tests
* #46951 [TensorExpr][CPU] Fix bool -> int casting
* #46950 Small changes/cleanup
* #46949 Force LLVM Compilation for CPU Tests

Differential Revision: [D24805246](https://our.internmc.facebook.com/intern/diff/D24805246)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/half_support.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#include <torch/csrc/jit/tensorexpr/ir.h>
+#include <torch/csrc/jit/tensorexpr/ir_printer.h>
+#include <torch/csrc/jit/tensorexpr/ir_visitor.h>
+#include <torch/csrc/jit/tensorexpr/tensor.h>
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+
+// Walk the Statment looking for Half size loads/stores.
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/half_support.h:10]: Probably best to do in a follow-up diff but it bugs me that something called "Checker" actually does mutation, and the doc comment doesn't give any indication of what it's doing either.  It'd be nice to rename and add a clearer comment.

File: torch/csrc/jit/tensorexpr/half_support.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    hasHalf_ = true;
+
+    return new Cast(Dtype(kFloat, child->dtype().lanes()), child);
+  }
+
+  Stmt* mutate(const Store* v) override {
+    const Expr* new_val = v->value()->accept_mutator(this);
+
+    if (v->value()->dtype().scalar_type() == ScalarType::Half) {
+      // TODO discards lanes.
```
Comment (bertmaher) [inline torch/csrc/jit/tensorexpr/half_support.h:32]: Hmmm is this `discard lanes` comment still true, or is it stale?  The line below seems to preserve lanes, but...?

**Review by bertmaher — APPROVED** [review]
Cool with landing this as-is since it's code motion, just a couple notes on this file in general.

File: torch/csrc/jit/tensorexpr/half_support.h (hunk 1 above)
Comment (eellison) [reply torch/csrc/jit/tensorexpr/half_support.h:10]: This comment (and your other comments) will be are addressed when I rebase onto  by Nick's PR https://github.com/pytorch/pytorch/pull/47448

Comment (nickgg) [reply torch/csrc/jit/tensorexpr/half_support.h:11]: Yeah this is fixed already :) (you're right though)

File: torch/csrc/jit/tensorexpr/half_support.h [hunk 3]
```diff
@@ -1,14 +1,15 @@
 #pragma once
 
-#include <torch/csrc/jit/codegen/fuser/cuda/resource_strings.h>
```
Comment (nickgg) [inline torch/csrc/jit/tensorexpr/half_support.h:3]: Cuda codegen needs this. You'll have to add it back there.

File: torch/csrc/jit/tensorexpr/half_support.h (hunk 2 above)
Comment (nickgg) [reply torch/csrc/jit/tensorexpr/half_support.h]: Should be fixed in master.

**Review by nickgg — DISMISSED** [review]
Let's make sure we don't break the Cuda backend since that's the bit thats turned on.

File: torch/csrc/jit/tensorexpr/half_support.h (hunk 3 above)
Comment (eellison) [reply torch/csrc/jit/tensorexpr/half_support.h:3]: This breaks the build when you're not using cuda... Why does it need to be in this file?

**Comment by eellison** [comment]
> Let's make sure we don't break the Cuda backend since that's the bit thats turned on.

Sure, Ci has to be green before i'm going to land

**Review by nickgg — APPROVED** [review]
🚢

**Pull request closed without merging**